File: app/src/main/java/ca/pkay/rcloneexplorer/RemoteConfig/RemoteConfig.java
package ca.pkay.rcloneexplorer.RemoteConfig;

import android.content.Context;
import android.os.Bundle;
import android.util.TypedValue;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

import org.json.JSONException;

import java.util.HashMap;

import ca.pkay.rcloneexplorer.R;
import ca.pkay.rcloneexplorer.Rclone;
import ca.pkay.rcloneexplorer.RuntimeConfiguration;
import ca.pkay.rcloneexplorer.rclone.Provider;
import ca.pkay.rcloneexplorer.util.ActivityHelper;
import es.dmoral.toasty.Toasty;

public class RemoteConfig extends AppCompatActivity implements RemotesConfigList.ProviderSelectedListener {


    public static final int CONFIG_EDIT_CODE = 139;

    public static final String CONFIG_EDIT_TARGET = "CONFIG_EDIT_TARGET";

    private final String OUTSTATE_TITLE = "ca.pkay.rcexplorer.remoteConfig.TITLE";
    private Fragment fragment;

    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(RuntimeConfiguration.attach(this, newBase));
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        applyTheme();
        setContentView(R.layout.activity_remote_config);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowHomeEnabled(true);
        }

        FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();

        if (savedInstanceState != null) {
            fragment = getSupportFragmentManager().findFragmentByTag("new config");
            if (fragment != null) {
                fragmentTransaction.replace(R.id.flFragment, fragment, "new config");
                fragmentTransaction.commit();

                String title = savedInstanceState.getString(OUTSTATE_TITLE);
                if (title != null) {
                    getSupportActionBar().setTitle(title);
                }
                return;
            }
        }
        fragment = RemotesConfigList.newInstance();
        fragmentTransaction.replace(R.id.flFragment, fragment, "config list");
        fragmentTransaction.commit();

        String shouldEdit = getIntent().getStringExtra(CONFIG_EDIT_TARGET);
        if(shouldEdit != null) {
            if(!shouldEdit.isEmpty()){
                Rclone rclone = new Rclone(this);
                HashMap<String, String> config = rclone.getConfig(getIntent().getStringExtra(CONFIG_EDIT_TARGET));

                if (config != null) {
                    try {
                        Provider provider = rclone.getProvider(config.get("type"));
                        fragment = new DynamicConfig(provider.getName(), config);
                        startConfig(provider);
                    } catch (JSONException e) {
                        throw new RuntimeException(e);
                    }
                }
            }
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        if (getSupportActionBar() != null) {
            CharSequence title = getSupportActionBar().getTitle();
            if (title != null) {
                outState.putString(OUTSTATE_TITLE, title.toString());
            }
        }
    }

    private void applyTheme() {
        ActivityHelper.applyTheme(this);
        TypedValue typedValue = new TypedValue();
        getTheme().resolveAttribute(R.attr.colorPrimaryDark, typedValue, true);
        getWindow().setStatusBarColor(typedValue.data);
    }

    @Override
    public boolean onSupportNavigateUp() {
        handleBackAction();
        return true;
    }

    @Override
    public void onBackPressed() {
        handleBackAction();
    }

    private void handleBackAction() {
        if (fragment instanceof RemotesConfigList) {
            finish();
        } else if (fragment instanceof DynamicConfig){
            if(((DynamicConfig) fragment).isEditConfig()){
                finish();
            } else {
                startProviderlist();
            }
        } else {
            startProviderlist();
        }
    }

    @Override
    public void onProviderSelected(Provider provider) {
        if (provider == null) {
            Toasty.error(this, getString(R.string.nothing_selected), Toast.LENGTH_SHORT, true).show();
            return;
        }

        switch (provider.getName()) {
            case "box":
            case "dropbox":
            case "pcloud":
            case "yandex":
            case "drive":
            case "google photos":
                fragment = new DynamicConfig(provider.getName(), true);
                break;
            default:
                fragment = new DynamicConfig(provider.getName());
                break;
        }
        startConfig(provider);
    }

    private void startConfig(Provider provider) {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.replace(R.id.flFragment, fragment, "new config");
        transaction.commit();

        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(provider.getNameCapitalized());
        }
    }

    private void startProviderlist() {
        fragment = RemotesConfigList.newInstance();
        FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
        fragmentTransaction.replace(R.id.flFragment, fragment, "config list");
        fragmentTransaction.commit();

        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(R.string.title_activity_remote_config);
        }
    }
}

File: app/src/main/java/ca/pkay/rcloneexplorer/RemoteConfig/RemoteConfig.kt
package ca.pkay.rcloneexplorer.RemoteConfig

import android.content.Context
import android.os.Bundle
import android.util.TypedValue
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.SearchView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import ca.pkay.rcloneexplorer.R
import ca.pkay.rcloneexplorer.Rclone
import ca.pkay.rcloneexplorer.RemoteConfig.ProviderListFragment.Companion.newProviderListConfig
import ca.pkay.rcloneexplorer.RemoteConfig.ProviderListFragment.ProviderSelectedListener
import ca.pkay.rcloneexplorer.RuntimeConfiguration
import ca.pkay.rcloneexplorer.rclone.Provider
import ca.pkay.rcloneexplorer.util.ActivityHelper
import es.dmoral.toasty.Toasty
import org.json.JSONException


class RemoteConfig : AppCompatActivity(), ProviderSelectedListener {

    private val OUTSTATE_TITLE = "ca.pkay.rcexplorer.remoteConfig.TITLE"
    private var mFragment: Fragment? = null

    companion object {
        const val CONFIG_EDIT_CODE = 139
        const val CONFIG_EDIT_TARGET = "CONFIG_EDIT_TARGET"
    }

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(RuntimeConfiguration.attach(this, newBase))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        applyTheme()
        setContentView(R.layout.activity_remote_config)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        val actionBar = supportActionBar
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true)
            actionBar.setDisplayShowHomeEnabled(true)
        }
        val fragmentTransaction = supportFragmentManager.beginTransaction()
        if (savedInstanceState != null) {
            mFragment = supportFragmentManager.findFragmentByTag("new config")
            fragmentTransaction.replace(R.id.flFragment, getCurrentFragment(), "new config")
            fragmentTransaction.commit()
            val title = savedInstanceState.getString(OUTSTATE_TITLE)
            if (title != null) {
                supportActionBar!!.setTitle(title)
            }
            return
        }

        mFragment = newProviderListConfig()
        fragmentTransaction.replace(R.id.flFragment, getCurrentFragment(), "config list")
        fragmentTransaction.commit()
        val shouldEdit = intent.getStringExtra(CONFIG_EDIT_TARGET)
        if (shouldEdit != null) {
            if (!shouldEdit.isEmpty()) {
                val rclone = Rclone(this)
                val config = rclone.getConfig(intent.getStringExtra(CONFIG_EDIT_TARGET))
                if (config != null) {
                    try {
                        val provider = rclone.getProvider(config["type"])
                        mFragment = DynamicRemoteConfigFragment(provider.name, config)
                        startConfig(provider)
                    } catch (e: JSONException) {
                        throw RuntimeException(e)
                    }
                }
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.remote_config_menu, menu)

        val searchbar = findViewById<SearchView>(R.id.search)

        searchbar.setOnCloseListener {
            toggleSearch(false)
            return@setOnCloseListener searchbar.isIconified
        }

        searchbar.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextChange(newText: String): Boolean {
                if(isProviderlist()) {
                    (getCurrentFragment() as ProviderListFragment).setSearchterm(newText)
                }
                if(isDynamicRemoteConfigFragment()) {
                    (getCurrentFragment() as DynamicRemoteConfigFragment).setSearchterm(newText)
                }
                return true
            }

            override fun onQueryTextSubmit(query: String): Boolean {
                return true
            }
        })
        return true
    }

    private fun toggleSearch(show: Boolean) {
        val searchbar = findViewById<SearchView>(R.id.search)
        if(show) {
            searchbar.visibility = View.VISIBLE
            searchbar.requestFocus()
            searchbar.isFocusable = true
            searchbar.isIconified = false
            searchbar.requestFocusFromTouch()
        } else {
            searchbar.visibility = View.GONE
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.app_bar_search -> {
                toggleSearch(true)
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (supportActionBar != null) {
            val title = supportActionBar!!.title
            if (title != null) {
                outState.putString(OUTSTATE_TITLE, title.toString())
            }
        }
    }
    override fun onSupportNavigateUp(): Boolean {
        handleBackAction()
        return true
    }

    override fun onBackPressed() {
        handleBackAction()
    }

    private fun handleBackAction() {
        toggleSearch(false)
        if (mFragment is ProviderListFragment) {
            finish()
        } else if (mFragment is DynamicRemoteConfigFragment) {
            if ((mFragment as DynamicRemoteConfigFragment).isEditConfig()) {
                finish()
            } else {
                var title: String? = null
                if (supportActionBar != null) {
                    title = supportActionBar!!.title.toString()
                }
                startProviderlist(title)
            }
        } else {
            startProviderlist()
        }
    }

    override fun onProviderSelected(provider: Provider) {
        if (provider == null) {
            Toasty.error(this, getString(R.string.nothing_selected), Toast.LENGTH_SHORT, true)
                .show()
            return
        }
        toggleSearch(false)
        mFragment = DynamicRemoteConfigFragment(provider.name)
        startConfig(provider)
    }

    private fun applyTheme() {
        ActivityHelper.applyTheme(this)
        val typedValue = TypedValue()
        theme.resolveAttribute(R.attr.colorPrimaryDark, typedValue, true)
        window.statusBarColor = typedValue.data
    }

    private fun startConfig(provider: Provider) {
        val transaction = supportFragmentManager.beginTransaction()
        transaction.replace(R.id.flFragment, getCurrentFragment(), "new config")
        transaction.commit()
        if (supportActionBar != null) {
            supportActionBar!!.title = provider.getNameCapitalized()
        }
    }

    private fun startProviderlist() {
        startProviderlist(null)
    }
    private fun startProviderlist(preselection: String?) {
        mFragment = newProviderListConfig(preselection)
        val fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction.replace(R.id.flFragment, getCurrentFragment(), "config list")
        fragmentTransaction.commit()
        if (supportActionBar != null) {
            supportActionBar!!.setTitle(R.string.title_activity_remote_config)
        }
    }

    private fun isProviderlist(): Boolean {
        if (mFragment is ProviderListFragment) {
            return true
        }
        return false
    }
    private fun isDynamicRemoteConfigFragment(): Boolean {
        if (mFragment is DynamicRemoteConfigFragment) {
            return true
        }
        return false
    }


    /**
     * This defauls to the provider list
     */
    private fun getCurrentFragment(): Fragment {
        if(mFragment == null) {
            return newProviderListConfig()
        }
        return mFragment!!
    }
}